front/components/assistant/conversation/ConversationTitle.tsx

```diff
@@ -39,16 +51,116 @@ export function ConversationTitle({
+      }
+      setIsEditingTitle(false);
+      setEditedTitle("");
+    } catch (e) {
+      alert("Failed to update title");
+    }
+  };
+
   if (isConversationLoading || isConversationError || !conversation) {
     return null;
   }
 
   return (
-    <div className="grid h-full max-w-full grid-cols-[1fr,auto] items-center gap-4">
-      <div className="overflow-hidden truncate">
-        <span className="font-bold">{conversation?.title || ""}</span>
-      </div>
+    <div className="grid h-full min-w-0 max-w-full grid-cols-[1fr,auto] items-center gap-4">
+      <div className="flex min-w-0 flex-row items-center gap-4">
+        {!isEditingTitle ? (
+          <div className="min-w-0 overflow-hidden truncate">
+            <span className="font-bold">{conversation?.title || ""}</span>
+          </div>
+        ) : (
+          <div className="flex-grow">
+            <input
+              className="w-full rounded-md font-bold"
+              value={editedTitle}
+              onChange={(e) => setEditedTitle(e.target.value)}
+              // this is to make sure the save button click
+              // is registered before the onBlur event
+              onFocus={() => (titleInputFocused.current = true)}
+              onBlur={() => {
+                setTimeout(() => {
+                  if (!saveButtonFocused.current) {
+                    setIsEditingTitle(false);
+                  }
+                  titleInputFocused.current = false;
+                }, 0);
+              }}
+              onKeyDown={(e) => {
+                if (e.key === "Enter") {
+                  return onTitleChange(editedTitle);
+                }
+              }}
+              autoFocus
+            />
+          </div>
+        )}
 
+        <div>
+          {isEditingTitle ? (
+            <div className="flex flex-row">
+              <div
+                onClick={(e: MouseEvent<HTMLDivElement>) => {
+                  e.preventDefault();
+                  return onTitleChange(editedTitle);
+                }}
+                // this is to make sure the save button click
+                // is registered before the onBlur event
+                onFocus={() => (saveButtonFocused.current = true)}
+                onBlur={() => {
+                  setTimeout(() => {
+                    if (!titleInputFocused.current) {
+                      setIsEditingTitle(false);
+                    }
+                    saveButtonFocused.current = false;
+                  }, 0);
+                }}
+              >
+                <IconButton icon={CheckIcon} />
+              </div>
+              <div>
+                <IconButton
+                  icon={XMarkIcon}
+                  onClick={() => {
+                    setIsEditingTitle(false);
+                    setEditedTitle("");
+                  }}
+                />
+              </div>
+            </div>
+          ) : (
+            <IconButton
```

Comment:
The alignment is not perfect here the icon is a bit above the text? Can we tweak it a bit?

Also that gray looks not super sharp maybe we can use the flavour that is darker?

Reply:
Fixed